### PR TITLE
Lightning API extensions

### PR DESCRIPTION
**Pull request by fiksn**

### CODE REVIEW

File: lightning/api_clnraw.go [hunk 1]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+	}
+
+	m := make(map[uint64]struct{})
+	for _, one := range outgoingChanIds {
+		m[one] = struct{}{}
+	}
+
+	for _, one := range chans {
+		_, ok := m[one.ChannelID]
+		if !ok {
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:868]: maybe avoid deep nesting with 
```
if ok {
    continue
}
...
```

File: lightning/api_clnraw.go [hunk 2]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+	if len(reply.TxID) == 0 {
+		return "", fmt.Errorf("invalid transaction id")
+	}
+
+	return reply.TxID, nil
+}
+
+// calculateExclusion is used since CLN can just exclude channels, so we tell it which channels we want and method calculates the "inverse" of it
+func (l *ClnRawLightningAPI) calculateExclusion(ctx context.Context, outgoingChanIds []uint64) ([]string, error) {
+	result := make([]string, 0)
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:844]: can we call it exclusions?

Comment (dcrystalj) [reply lightning/api_clnraw.go:844]: I would move this to line 865 so related things are close together

File: lightning/api_clnraw.go [hunk 3]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+
+	if len(reply.TxID) == 0 {
+		return "", fmt.Errorf("invalid transaction id")
+	}
+
+	return reply.TxID, nil
+}
+
+// calculateExclusion is used since CLN can just exclude channels, so we tell it which channels we want and method calculates the "inverse" of it
+func (l *ClnRawLightningAPI) calculateExclusion(ctx context.Context, outgoingChanIds []uint64) ([]string, error) {
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:843]: Its more natural to put helper function under (after) calling one.

File: lightning/api_clnraw.go [hunk 4]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+
+			result = append(result, fmt.Sprintf("%s/%d", id, direction))
+		}
+	}
+
+	return result, nil
+}
+
+// PayInvoice - API call.
+func (l *ClnRawLightningAPI) PayInvoice(ctx context.Context, paymentRequest string, sats int64, outgoingChanIds []uint64) (*PaymentResp, error) {
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:897]: looks like this are `msats` and not `sats` if we need to multiply with 1000 ?

File: lightning/api_clnraw.go [hunk 5]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+
+// GetPaymentStatus - API call.
+func (l *ClnRawLightningAPI) GetPaymentStatus(ctx context.Context, paymentHash string) (*PaymentResp, error) {
+	var reply ClnPaymentEntries
+	err := l.connection.Call(ctx, ListPayments, []interface{}{nil, paymentHash}, &reply)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(reply.Entries) != 1 {
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:943]: In what case can we get 2 entries here ? Looks a bit dangerous to say it is invalid

File: lightning/api_clnraw.go [hunk 6]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+		return nil, fmt.Errorf("invalid response")
+	}
+
+	hash := reply.Entries[0].PaymentHash
+
+	if hash != paymentHash {
```
Comment (dcrystalj) [inline lightning/api_clnraw.go:949]: nice. totally not trusting their api :)

File: lightning/api_clnsocket_test.go [hunk 7]
```diff
@@ -340,3 +340,316 @@ func TestClnGetPaymentsRaw(t *testing.T) {
...
+	defer cancel()
+
+	err := api.ConnectPeer(ctx, fmt.Sprintf("%s@%s", pubkey, host))
+	assert.NoError(t, err)
+}
+
+func TestClnGetOnChainAddress(t *testing.T) {
+	data := clnData(t, "cln_newaddr")
+
+	_, api, closer := clnCommon(t, func(c net.Conn) {
```
Comment (dcrystalj) [inline lightning/api_clnsocket_test.go:383]: repeated part in multiple tests

File: lightning/api_grpc.go [hunk 8]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout")
+		default:
```
Comment (dcrystalj) [inline lightning/api_grpc.go:900]: I really dont understand this pattern

looks like you dont need select

```
 if ctx.Err() != nil {
    return nil, fmt.Errorf("timeout")
 }
 ```

File: lightning/api_grpc.go [hunk 9]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+			// TODO: we don't know the hash here, else we could return l.GetPaymentStatus()
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout")
```
Comment (dcrystalj) [inline lightning/api_grpc.go:899]: I think you could just reutrn ctx.Err() ?

File: lightning/api_grpc.go [hunk 10]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+		default:
+			// Do nothing
+		}
+
+		event, err := resp.Recv()
+
```
Comment (dcrystalj) [inline lightning/api_grpc.go:905]: remove empty line

File: lightning/api_grpc.go [hunk 11]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+				Status:   Pending,
+			}, nil
+		}
+	}
+}
+
+// CreateInvoice API.
+func (l *LndGrpcLightningAPI) CreateInvoice(ctx context.Context, sats int64, preimage string, memo string, expiry time.Duration) (*InvoiceResp, error) {
+	var err error
+	req := &lnrpc.Invoice{}
```
Comment (dcrystalj) [inline lightning/api_grpc.go:994]: I dont like setting everything to zero value and then in next 2 lines change something again

File: lightning/api_grpc.go [hunk 12]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	if sats > 0 {
+		req.Value = sats
+	}
+
+	resp, err := l.Client.AddInvoice(ctx, req)
+
```
Comment (dcrystalj) [inline lightning/api_grpc.go:1009]: remove

File: lightning/api_grpc_test.go [hunk 13]
```diff
@@ -314,3 +328,280 @@ func TestGetChanInfoGrpc(t *testing.T) {
...
+
+		mr.
+			EXPECT().
+			SendPaymentV2(gomock.Any(), gomock.Any()).
+			Return(c, nil)
+
+		resp, err := api.PayInvoice(context.Background(), "lnbcrt13370n1p3lwhhnpp5zfvpdpwp77wmgpyawatm550uv9cvx0hv9hgukxd6u0pqqdhcdymsdqqcqzpgxqyz5vqsp5xsum9s4cpkvw27f6smk4daxqkpjgmah8xxhs8ty34fm4srmwfvjs9qyyssqlx5zk7j8lfeelzmpsk0mmwp3583tl52j8us2q9nt05vrmtp3sasxzc8wyjchrum67sllzr52gjz26rcrye6y4vrlpr6pyv9jhhlrp2cq52rxf5", 0, []uint64{128642860515328})
+		ctrl.Finish()
+
+		one.Second(resp, err)
```
Comment (dcrystalj) [inline lightning/api_grpc_test.go:510]: strange wording. one-second

File: lightning/rest.go [hunk 14]
```diff
@@ -292,15 +289,51 @@ func (h *HTTPAPI) HTTPListPayments(ctx context.Context, req *http.Request, input
...
+func (h *HTTPAPI) HTTPPeers(ctx context.Context, req *http.Request, input *ConnectPeerRequestOverride) error {
+	req = req.WithContext(ctx)
+	req.Method = http.MethodPost
+
+	u, err := url.Parse(fmt.Sprintf("%s/v1/peers", req.URL))
+	if err != nil {
+		return fmt.Errorf("invalid url %s", err)
+	}
+
+	s, _ := json.Marshal(input)
```
Comment (dcrystalj) [inline lightning/rest.go:302]: ignorring error ?

File: lightning/rest.go [hunk 15]
```diff
@@ -292,15 +289,51 @@ func (h *HTTPAPI) HTTPListPayments(ctx context.Context, req *http.Request, input
...
+	req.Body = io.NopCloser(b)
+
+	resp, err := h.Do(req)
+	if err != nil {
+		return fmt.Errorf("http request failed %s", err)
+	}
+
+	defer resp.Body.Close()
+
+	alreadyConnected := false
```
Comment (dcrystalj) [inline lightning/rest.go:315]: hard to follow. It should be defined closer to its usage

File: lightning/rest.go [hunk 16]
```diff
@@ -292,15 +289,51 @@ func (h *HTTPAPI) HTTPListPayments(ctx context.Context, req *http.Request, input
...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	if strings.Contains(string(body), "already connected to peer") {
```
Comment (dcrystalj) [inline lightning/rest.go:320]: I suggest you replace it with definition directly or just but it all in one if
```
isAlreadyConnected := strings.Contains(string(body), "already connected to peer")
if resp.StatusCode != http.StatusOK && !isAlreadyConnected {
    return fmt.Errorf("http got error %d %s", resp.StatusCode, body)
}
```

File: lightning/rest.go [hunk 17]
```diff
@@ -292,15 +289,51 @@ func (h *HTTPAPI) HTTPListPayments(ctx context.Context, req *http.Request, input
 	return &data, nil
 }
 
-// GetHTTPRequest - generic method for doing HTTP requests
-func (h *HTTPAPI) GetHTTPRequest(getData GetDataCall) (*http.Request, *http.Transport, error) {
+// HTTPPeers - invokes peers method.
+func (h *HTTPAPI) HTTPPeers(ctx context.Context, req *http.Request, input *ConnectPeerRequestOverride) error {
+	req = req.WithContext(ctx)
```
Comment (dcrystalj) [inline lightning/rest.go:294]: consider creating common function for Post requests. Most of this one is copy/pasted 6 times now

File: lightning/api_grpc.go [hunk 18]
```diff
@@ -790,7 +783,256 @@ func (l *LndGrpcLightningAPI) SubscribeFailedForwards(ctx context.Context, outch
...
+	}
+
+	_, err := l.Client.ConnectPeer(ctx, &lnrpc.ConnectPeerRequest{
+		Addr:    &lnrpc.LightningAddress{Host: split[1], Pubkey: split[0]},
+		Perm:    false,
+		Timeout: 10,
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "already connected to peer") {
+			err = nil
```
[the remark this answers is not in the document]
Comment (fiksn) [reply lightning/api_grpc.go:805]: Oh right, I think I wanted to do something more but then removed it.

File: lightning/api_clnraw.go [hunk 19]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+	)
+
+	var fundsReply ClnFundsChainResp
+	err := l.connection.Call(ctx, ListFunds, []string{}, &fundsReply)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &Funds{}
+	f.ConfirmedBalance = 0
```
[the remark this answers is not in the document]
Comment (fiksn) [reply lightning/api_clnraw.go:792]: I don't completely agree here. Yeah, structs are initialized with everything zeroed out but this is an implementation detail that you shouldn't depend on imo. 

What I mean to say if there was `f := NewFunds()` it'd be in our domain and we wouldn't need to do the "redundant" assignment (altho I am quite sure compiler will optimize it out). But in this case this creates an implicit "dependency" on go working like it does and our code. Of course such a (breaking) change will likely never happen but still.

File: lightning/api_clnraw.go [hunk 20]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+	var reply ClnWithdrawResp
+
+	feerate := "normal"
+	switch urgency {
+	case Urgent:
+		feerate = "urgent"
+	case Normal:
+		feerate = "normal"
+	case Low:
+		feerate = "slow" // slow not low, beware!
```
[the remark this answers is not in the document]
Comment (fiksn) [reply lightning/api_clnraw.go:822]: The comment is just because "low" and "slow" are so common and somebody might think the feerate string is a typo for "low". CLN allows you to pass it "slow" string as parameter for feerate.

And we have an abstraction {Urgent, Normal, Low} "urgency". Possibly this naming needs to be improved.

File: lightning/api_clnraw.go [hunk 21]
```diff
@@ -767,3 +742,288 @@ func (l *ClnRawLightningAPI) GetInternalChannelsAll(ctx context.Context) (map[st
...
+	}
+
+	minConf := 1
```
[the remark this answers is not in the document]
Comment (fiksn) [reply lightning/api_clnraw.go:825]: minConf is the number of confirmations your UTXOs needs to have to be used. I don't think it makes sense to use something > 1 since this is for sending. It makes sense to distinguish between confirmed balance and unconfirmed balance. If you want to use the second it is more likely to fail. But if you use a confirmed UTXO.. well there could be a reorg and your balance will change and tx fail (but this is more of a rare corner-case) and tx might fail for other reasons too.

File: lightning/api_clnraw.go (hunk 4 above)
Comment (fiksn) [reply lightning/api_clnraw.go:897]: Parameter name is correct, we get "sats" but need to multiply by 1000 since CLN API uses msats.

Comment (fiksn) [reply lightning/api_clnraw.go:897]: Actually this got me thinking, probably an API with milisats is more future proof when 1 sat = 1 cent 😄

File: lightning/api_clnraw.go (hunk 5 above)
Comment (fiksn) [reply lightning/api_clnraw.go:943]: We sort of call get with an unique id (except that method name is list). So if we get 0 entries the thing doesn't exist but if we get more than 1 something is really screwed up. Maybe we should split the cases for more verbose errors but in any case if the result does not have exactly one entry the function must fail.

File: lightning/api_grpc.go (hunk 8 above)
Comment (fiksn) [reply lightning/api_grpc.go:900]:
```
// If Done is not yet closed, Err returns nil.
// If Done is closed, Err returns a non-nil error explaining why:
// Canceled if the context was canceled
// or DeadlineExceeded if the context's deadline passed.
// After Err returns a non-nil error, successive calls to Err return the same error.
```

Looks like you're right 😊  I just invented an ugly antipattern, let's fix this.

File: lightning/rest.go (hunk 17 above)
Comment (fiksn) [reply lightning/rest.go:291]: 👍 Actually I know why I avoided that before, without generics you can't really make the marshaller work fine, if you pass in an interface it doesn't work as expected. Fixed that now.

**Review by dcrystalj — APPROVED** [review]
